core/src/main/scala/org/apache/spark/sql/execution/ExistingPlans.scala

```diff
@@ -233,13 +242,17 @@ private[sql] final case class ZipPartitionScan(basePlan: CodegenSupport,
   private val consumedVars: ArrayBuffer[ExprCode] = ArrayBuffer.empty
   private val inputCode = basePlan.asInstanceOf[CodegenSupport]
 
-  override def children: Seq[SparkPlan] = basePlan :: otherPlan :: Nil
+  private val withShuffle = ShuffleExchange(HashPartitioning(
+    ClusteredDistribution(otherPartKeys)
+        .clustering, inputCode.inputRDDs().head.getNumPartitions), otherPlan)
+
+  override def children: Seq[SparkPlan] = basePlan :: withShuffle :: Nil
```

Comment:
Why is this always a shuffle?? What if the partitioning of the two match (which is the more common case)? Either way we discussed that this will be handled automatically by EnsureRequirements and this should not be done explicitly. I guess you will need this if the plan is being used after EnsureRequirements but then it should use a similar way of determining whether to do shuffle, broadcast or collocate join like done by EnsureRequirements.